File: src/lib/business-date.ts
/** 中国标准时间相对 UTC 的固定偏移量。 */
const SHANGHAI_OFFSET_MS = 8 * 60 * 60 * 1000

/** 一天的毫秒数。 */
const DAY_MS = 24 * 60 * 60 * 1000

/** 系统默认打卡话题的稳定 slug。 */
export const CHECK_IN_TOPIC_SLUG = 'check-in-challenge'

/**
 * 将数字补齐为两位字符串。
 * @param value 月或日数值
 * @returns 两位十进制字符串
 */
function padTwo(value: number): string {
  return value < 10 ? '0' + String(value) : String(value)
}

/**
 * 将任意时间转换为 Asia/Shanghai 的 YYYY-MM-DD 业务日期。
 * 中国标准时间全年固定为 UTC+8，因此使用 UTC 字段读取偏移后的时间，
 * 不受 Node 进程或数据库服务器本地时区影响。
 * @param date 待转换时间
 * @returns 北京时间业务日期
 */
export function toShanghaiDateKey(date: Date = new Date()): string {
  const shifted = new Date(date.getTime() + SHANGHAI_OFFSET_MS)
  return String(shifted.getUTCFullYear())
    + '-'
    + padTwo(shifted.getUTCMonth() + 1)
    + '-'
    + padTwo(shifted.getUTCDate())
}

/**
 * 获取指定时间前一天对应的北京时间业务日期。
 * @param date 参照时间
 * @returns 前一天的 YYYY-MM-DD 日期
 */
export function getPreviousShanghaiDateKey(date: Date = new Date()): string {
  return toShanghaiDateKey(new Date(date.getTime() - DAY_MS))
}

/**
 * 计算当前仍然有效的连续打卡天数。
 * 最后打卡发生在今天或昨天时连续记录仍有效；更早则视为已经断签。
 * @param streakDays 数据库保存的连续天数
 * @param lastCheckInAt 最后打卡时间
 * @param now 当前时间，测试时可传入固定值
 * @returns 对外展示的有效连续天数
 */
export function getEffectiveStreakDays(
  streakDays: number,
  lastCheckInAt: Date | null,
  now: Date = new Date()
): number {
  if (streakDays <= 0 || !lastCheckInAt) {
    return 0
  }

  const lastDateKey = toShanghaiDateKey(lastCheckInAt)
  const todayDateKey = toShanghaiDateKey(now)
  const yesterdayDateKey = getPreviousShanghaiDateKey(now)

  return lastDateKey === todayDateKey || lastDateKey === yesterdayDateKey
    ? streakDays
    : 0
}

File: internal/pkg/businessdate/businessdate.go
// Package businessdate 提供 Asia/Shanghai 业务日期相关工具。
// 打卡连续天数的判断依赖本包，避免服务器时区影响业务日期。
package businessdate

import "time"

// CheckInTopicSlug 是系统默认打卡话题的稳定 slug。
const CheckInTopicSlug = "check-in-challenge"

// shanghaiOffset 是中国标准时间相对 UTC 的固定偏移（+8 小时）。
const shanghaiOffset = 8 * 60 * 60

// daySeconds 是一天的秒数。
const daySeconds = 24 * 60 * 60

// padTwo 将数字补齐为两位字符串。
func padTwo(value int) string {
	if value < 10 {
		return "0" + itoa(value)
	}
	return itoa(value)
}

// itoa 是轻量整数转字符串实现，避免引入 strconv 依赖。
func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	negative := false
	if value < 0 {
		negative = true
		value = -value
	}
	buf := [20]byte{}
	pos := len(buf)
	for value > 0 {
		pos--
		buf[pos] = byte('0' + value%10)
		value /= 10
	}
	if negative {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}

// ToShanghaiDateKey 将任意时间转换为 Asia/Shanghai 的 YYYY-MM-DD 业务日期。
// 使用固定 UTC+8 偏移，不受进程或数据库服务器本地时区影响。
func ToShanghaiDateKey(t time.Time) string {
	shifted := t.UTC().Add(time.Duration(shanghaiOffset) * time.Second)
	return itoa(shifted.Year()) + "-" + padTwo(int(shifted.Month())) + "-" + padTwo(shifted.Day())
}

// GetPreviousShanghaiDateKey 获取指定时间前一天对应的北京时间业务日期。
func GetPreviousShanghaiDateKey(t time.Time) string {
	return ToShanghaiDateKey(t.Add(-time.Duration(daySeconds) * time.Second))
}

// GetEffectiveStreakDays 计算当前仍然有效的连续打卡天数。
// 最后打卡发生在今天或昨天时连续记录仍有效；更早则视为已经断签。
func GetEffectiveStreakDays(streakDays int, lastCheckInAt *time.Time, now time.Time) int {
	if streakDays <= 0 || lastCheckInAt == nil {
		return 0
	}
	lastDateKey := ToShanghaiDateKey(*lastCheckInAt)
	todayDateKey := ToShanghaiDateKey(now)
	yesterdayDateKey := GetPreviousShanghaiDateKey(now)
	if lastDateKey == todayDateKey || lastDateKey == yesterdayDateKey {
		return streakDays
	}
	return 0
}
